### PR TITLE
Move geo code from parser to geo package.

### PR DESCRIPTION
**Pull request by gfr10598**
This PR splits the content of the original metadata.go into two subsets,
part that is independent (in geo/) and part that is dependent on other parser and schema code.  This was done by cutting and pasting chunks of code from metadata.go into geo/annotation.go (and corresponding _test files), to ensure nothing was overlooked or duplicated.

Then, files are renamed, to better reflect that they handle annotation functions.
parser/
  metadata.go -> annotation.go
  metadata_test.go -> annotation.go

Clean up, using appropriate prefixes in appropriate places.

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/etl/255)

### CODE REVIEW

**Comment by yachang** [comment]
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[geo/annotation.go, line 1 at r3](https://reviewable.io:443/reviews/m-lab/etl/255#-KvSuLOHXj54sj22vjo6:-KvSuLOHXj54sj22vjo7:b-jmuwt3) ([raw file](https://github.com/m-lab/etl/blob/c95d3f6e84dd13997b0a616349a8cb073227b6dd/geo/annotation.go#L1)):*
> ```Go
> package geo
> ```

My point of view:

name the package "annotation", so is dir name.

When I add IP-ASN code in the near future, they will be in the same package with GeoLocation code.

All Geolocation code can be named starting with "geo".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/255)*

**Comment by yachang** [comment]
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[annotation/geo.go, line 30 at r4](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTQTCN4r_TfpGMNn1d:-KvTQTCN4r_TfpGMNn1e:b-5vvssr) ([raw file](https://github.com/m-lab/etl/blob/0031bb69818c91460735648da476396dc1477c70/annotation/geo.go#L30)):*
> ```Go
> }
> 
> func checkFlags() {
> ```

This func actually set the value for var IPAnnotationEnabled by checking flag value.

setValue() might be a better name 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/255)*

**Comment by gfr10598** [comment]
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[annotation/geo.go, line 1 at r3](https://reviewable.io:443/reviews/m-lab/etl/255#-KvSuLOHXj54sj22vjo6:-KvTRWo5UEhLkvTyw7UG:b-dwmq42) ([raw file](https://github.com/m-lab/etl/blob/c95d3f6e84dd13997b0a616349a8cb073227b6dd/annotation/geo.go#L1)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

My point of view:

name the package "annotation", so is dir name.

When I add IP-ASN code in the near future, they will be in the same package with GeoLocation code.

All Geolocation code can be named starting with "geo".


</blockquote></details>

This is a left-over.  Not showing up in my client, so I'm not sure why it shows up here.

---

*[annotation/geo.go, line 30 at r4](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTQTCN4r_TfpGMNn1d:-KvTRNSDMnCEUYGowWCR:bkvaw8m) ([raw file](https://github.com/m-lab/etl/blob/0031bb69818c91460735648da476396dc1477c70/annotation/geo.go#L30)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

This func actually set the value for var IPAnnotationEnabled by checking flag value.

setValue() might be a better name 
</blockquote></details>

I'm changing to getFlagValues, which I think conveys it better.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/255)*

**Comment by yachang** [comment]
Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[annotation/geo.go, line 72 at r4](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTRpw5E5jLChNVkSal:-KvTRpw5E5jLChNVkSam:brslhyn) ([raw file](https://github.com/m-lab/etl/blob/0031bb69818c91460735648da476396dc1477c70/annotation/geo.go#L72)):*
> ```Go
> 		reqData = append(reqData, RequestData{ip, 0, timestamp})
> 	}
> 	annotationData := GetBatchMetaData(BatchURL, reqData)
> ```

Change the func name to GetBatchAnnotationData() ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/255)*

File: annotation/geo.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+	if annotationData != nil && annotationData.Geo != nil {
+		*geo = *annotationData.Geo
+	}
+}
+
+// GetMetaData combines the functionality of QueryAnnotationService and
+// ParseJSONMetaDataResponse to query the annotator service and return
+// the corresponding MetaData if it can, or a nil pointer if it
+// encounters any error and cannot get the data for any reason
+func GetMetaData(url string) *MetaData {
```
Comment (yachang) [inline annotation/geo.go]: Change the func name to GetBatchAnnotationData() ?

File: annotation/geo.go [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+	annotatorResponse, err := QueryAnnotationService(url)
+	if err != nil {
+		metrics.AnnotationErrorCount.
+			With(prometheus.Labels{"source": "Error querying annotation service"}).Inc()
+		log.Println(err)
+		return nil
+	}
+
+	// Safely parse the JSON response and pass it back to the caller
+	metaDataFromResponse, err := ParseJSONMetaDataResponse(annotatorResponse)
```
Comment (yachang) [inline annotation/geo.go]: change var name "metaDataFromResponse" as well

File: annotation/geo.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+	// Copy response into a byte slice
+	return ioutil.ReadAll(resp.Body)
+}
+
+// ParseJSONMetaDataResponse takes a byte slice containing the test of
+// the JSON from the annotator service and parses it into a MetaData
+// struct, for easy manipulation. It returns a pointer to the struct on
+// success and an error if an error occurs.
+func ParseJSONMetaDataResponse(jsonBuffer []byte) (*MetaData, error) {
```
Comment (yachang) [inline annotation/geo.go]: remove "MetaData" from func name

File: annotation/geo.go [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+// BatchQueryAnnotationService takes a url to POST the request to and
+// a slice of RequestDatas to be sent in the body in a JSON
+// format. It will copy the response into a []byte and return it to
+// the user, returning an error if any occurs
+func BatchQueryAnnotationService(url string, data []RequestData) ([]byte, error) {
+	encodedData, err := json.Marshal(data)
+	if err != nil {
+		metrics.AnnotationErrorCount.
```
Comment (yachang) [inline annotation/geo.go]: There is only AnnotationErrorCount. We need to count how many annotation requests as well.

**Comment by gfr10598** [comment]
Review status: 0 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

*[annotation/geo.go, line 72 at r4](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTRpw5E5jLChNVkSal:-KvTYo8yOhZHullPw1h5:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/0031bb69818c91460735648da476396dc1477c70/annotation/geo.go#L72)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

Change the func name to GetBatchAnnotationData() ?
</blockquote></details>

Done.

---

*[annotation/geo.go, line 104 at r5](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTAWnG8LrQ-3wAxARO-r5-104:-KvTYnquhgBdnHmyrkqL:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/952782449720e244d37bcb03231d54769a77dde1/annotation/geo.go#L104)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

Change the func name to GetBatchAnnotationData() ?
</blockquote></details>

Done.

---

*[annotation/geo.go, line 115 at r5](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTAWnG8LrQ-3wAxARO-r5-115:-KvTZ4UNn_X52yErjK_p:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/952782449720e244d37bcb03231d54769a77dde1/annotation/geo.go#L115)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

change var name "metaDataFromResponse" as well
</blockquote></details>

Done.

---

*[annotation/geo.go, line 155 at r5](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTAWnG8LrQ-3wAxARO-r5-155:-KvTZ7m5TV1yIn36ltS5:b-jkl7yx) ([raw file](https://github.com/m-lab/etl/blob/952782449720e244d37bcb03231d54769a77dde1/annotation/geo.go#L155)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

remove "MetaData" from func name 
</blockquote></details>

Changed to GeoData.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/255)*

File: parser/annotation.go [hunk 5]
```diff
@@ -0,0 +1,287 @@
+package parser
```
Comment (yachang) [inline parser/annotation.go]: rename this file to geo_annotation.go

**Comment by gfr10598** [comment]
Review status: 0 of 7 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[annotation/geo.go, line 198 at r5](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTAWnG8LrQ-3wAxARO-r5-198:-KvTax77qDHg80noJT5x:brd9f6z) ([raw file](https://github.com/m-lab/etl/blob/952782449720e244d37bcb03231d54769a77dde1/annotation/geo.go#L198)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

There is only AnnotationErrorCount. We need to count how many annotation requests as well.
</blockquote></details>

Done.  This also exposed some overcounting of errors.  Left some TODOs to fix some, but already fixed others.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/255)*

**Comment by gfr10598** [comment]
Review status: 0 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

*[parser/annotation.go, line 1 at r7](https://reviewable.io:443/reviews/m-lab/etl/255#-KvIS-ub7czDtbqheubN-r7-1:-KvTbRYO8JRjS8IYWr-m:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/6514ffd771f1f9b5a78587006c1c5913c393b392/parser/annotation.go#L1)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

rename this file to geo_annotation.go
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/255)*

File: annotation/geo.go [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+		}
+		*geoDest[index] = *data.Geo
+
+	}
+}
+
+// GetAndInsertGeolocationIPStruct takes a NON-NIL pointer to a
+// pre-allocated GeolocationIP struct, an IP address, and a
+// timestamp. It will connect to the annotation service, get the
+// metadata, and insert the metadata into the reigion pointed to by
```
Comment (yachang) [inline annotation/geo.go]: Update the comments, get rid of metadata here.

File: annotation/geo_structs.go [hunk 7]
```diff
@@ -30,7 +30,7 @@ type IPASNData struct{}
 // The main struct for the metadata, which holds pointers to the
```
Comment (yachang) [inline annotation/geo_structs.go]: Update the doc

File: metrics/metrics.go [hunk 8]
```diff
@@ -56,14 +57,24 @@ var (
...
+
 	// Measures the number of annotation errors
 	// Provides metrics:
 	//    etl_annotator_Error_Count
 	// Example usage:
 	//    metrics.AnnotationErrorCount.Inc() / .Dec()
 	AnnotationErrorCount = prometheus.NewGaugeVec(prometheus.GaugeOpts{
 		Name: "etl_annotator_Error_Count",
-		Help: "The current number of errors encountered while attemting to add metadata.",
+		Help: "The current number of errors encountered while attempting to add metadata.",
```
Comment (yachang) [inline metrics/metrics.go]: Update the Help string as well.

File: parser/geo_annotation.go [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+
+	"github.com/m-lab/etl/annotation"
+	"github.com/m-lab/etl/metrics"
+	"github.com/m-lab/etl/schema"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// AddGeoDataSSConnSpec takes a pointer to a
+// Web100ConnectionSpecification struct and a timestamp. With these,
+// it will fetch the appropriate metadata and add it to the hop struct
```
Comment (yachang) [inline parser/geo_annotation.go]: ditto

File: parser/geo_annotation.go [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+	}(timerStart)
+
+	ipSlice := []string{spec.Local_ip, spec.Remote_ip}
+	geoSlice := []*annotation.GeolocationIP{&spec.Local_geolocation, &spec.Remote_geolocation}
+	annotation.FetchGeoAnnotations(ipSlice, timestamp, geoSlice)
+}
+
+// AddGeoDataPTConnSpec takes a pointer to a
+// MLabConnectionSpecification struct and a timestamp. With these, it
+// will fetch the appropriate metadata and add it to the hop struct
```
Comment (yachang) [inline parser/geo_annotation.go]: ditto

File: parser/geo_annotation.go [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+			Observe(float64(time.Since(tStart).Nanoseconds()))
+	}(timerStart)
+	requestSlice := CreateRequestDataFromPTHops(hops, timestamp)
+	annotationData := annotation.GetBatchGeoData(annotation.BatchURL, requestSlice)
+	AnnotatePTHops(hops, annotationData, timestamp)
+}
+
+// AnnotatePTHops takes a slice of hop pointers, the annotation data
+// mapping ip addresses to metadata and a timestamp. It will then use
+// these to attach the appropriate metadata to the PT hops.
```
Comment (yachang) [inline parser/geo_annotation.go]: Update doc

File: parser/geo_annotation.go [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+		annotation.GetAndInsertGeolocationIPStruct(&hop.Dest_geolocation, hop.Dest_ip, timestamp)
+	} else {
+		metrics.AnnotationErrorCount.With(prometheus.
+			Labels{"source": "PT Hop had no dest_ip!"}).Inc()
+	}
+}
+
+// AddGeoDataNDTConnSpec takes a connection spec and a timestamp and
+// annotates the connection spec with metadata associated with each IP
+// Address. It will either sucessfully add the metadata or fail
```
Comment (yachang) [inline parser/geo_annotation.go]: ditto

**Comment by gfr10598** [comment]
Review status: 0 of 8 files reviewed at latest revision, 15 unresolved discussions.

---

*[annotation/geo.go, line 91 at r9](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTAWnG8LrQ-3wAxARO-r9-91:-KvTf-YNd-wxpdWVGBW4:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/9405fa9932600637ca3116e56f90d5bd8603248d/annotation/geo.go#L91)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

Update the comments, get rid of metadata here.
</blockquote></details>

Done.

---

*[annotation/geo_structs.go, line 30 at r9](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTAWnG8LrQ-3wAxARP-r9-30:-KvTffM18L86qIC-Dr0G:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/9405fa9932600637ca3116e56f90d5bd8603248d/annotation/geo_structs.go#L30)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

Update the doc
</blockquote></details>

Done.

---

*[metrics/metrics.go, line 77 at r9](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTbDh2-F17e1ZGCQTn-r9-77:-KvTfgUzgNxhxrLJxAfT:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/9405fa9932600637ca3116e56f90d5bd8603248d/metrics/metrics.go#L77)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

Update the Help string as well.
</blockquote></details>

Done.

---

*[parser/geo_annotation.go, line 1 at r7](https://reviewable.io:443/reviews/m-lab/etl/255#-KvIS-ub7czDtbqheubN-r7-1:-KvTfiuhSy1ugAC3rAd3:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/6514ffd771f1f9b5a78587006c1c5913c393b392/parser/geo_annotation.go#L1)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[parser/geo_annotation.go, line 20 at r9](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTbRpmowEz9V0qcZcS-r9-20:-KvTfka9uc8DaE25kTCI:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/9405fa9932600637ca3116e56f90d5bd8603248d/parser/geo_annotation.go#L20)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[parser/geo_annotation.go, line 43 at r9](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTbRpmowEz9V0qcZcS-r9-43:-KvTflABpaPk9irLrIM1:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/9405fa9932600637ca3116e56f90d5bd8603248d/parser/geo_annotation.go#L43)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[parser/geo_annotation.go, line 81 at r9](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTbRpmowEz9V0qcZcS-r9-81:-KvTflnKrbx9ObZ1gH-d:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/9405fa9932600637ca3116e56f90d5bd8603248d/parser/geo_annotation.go#L81)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

Update doc
</blockquote></details>

Done.

---

*[parser/geo_annotation.go, line 174 at r9](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTbRpmowEz9V0qcZcS-r9-174:-KvTfmvIGxrrxsTmUvG7:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/9405fa9932600637ca3116e56f90d5bd8603248d/parser/geo_annotation.go#L174)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/255)*

**Comment by yachang** [comment]
Review status: 0 of 8 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[metrics/metrics.go, line 64 at r11](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTgwXM_2MJIALHrYaW:-KvTgwXM_2MJIALHrYaX:b-mkklpz) ([raw file](https://github.com/m-lab/etl/blob/1d530c23b7a294a224d1eab6a711957379da73e8/metrics/metrics.go#L64)):*
> ```Go
> 	//    etl_annotator_Request_Count
> 	// Example usage:
> 	//    metrics.AnnotationRequestCount.Inc() / .Dec()
> ```

No Dec() needed for this counter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/255)*

File: metrics/metrics.go [hunk 13]
```diff
@@ -56,14 +57,24 @@ var (
 		Help: "The total time to annotate, in nanoseconds.",
 	}, []string{"test_type"})
 
+	// Measures the number of annotation requests
+	// Provides metrics:
+	//    etl_annotator_Request_Count
+	// Example usage:
+	//    metrics.AnnotationRequestCount.Inc() / .Dec()
```
Comment (yachang) [inline metrics/metrics.go]: No Dec() needed for this counter.

**Comment by yachang** [comment]
LGTM

**Comment by gfr10598** [comment]
Review status: 0 of 8 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[metrics/metrics.go, line 64 at r11](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTbDh2-F17e1ZGCQTn-r11-64:-KvThQHAEwWOKsE0Li3h:b97gzza) ([raw file](https://github.com/m-lab/etl/blob/1d530c23b7a294a224d1eab6a711957379da73e8/metrics/metrics.go#L64)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

No Dec() needed for this counter.
</blockquote></details>

Will fix in next PR.

---

*[metrics/metrics.go, line 64 at r11](https://reviewable.io:443/reviews/m-lab/etl/255#-KvTgwXM_2MJIALHrYaW:-KvThSP6KaVBKlIfRfqq:b-rqxvp7) ([raw file](https://github.com/m-lab/etl/blob/1d530c23b7a294a224d1eab6a711957379da73e8/metrics/metrics.go#L64)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

No Dec() needed for this counter.
</blockquote></details>

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/255)*

**Pull request merged**